designs/memory-management-design.md

```diff
@@ -0,0 +1,275 @@
+```
+assert number of references of o is 1
+assert size of o(in bytes) is 4
+assert type of the fields in o is [value]
+```
+
+### 6. Access is not assignment
+
+**Case:**
+```
+class Rat(object):
+    y: int = 0
+    def __init__(self: Rat):
+		self.y = 1
+
+x: Rat = None
+x = Rat()
+x.y
+print(x.y)
+```
+**Expected:**
+Let `o` be the object referred to by variable `x`
+```
+assert number of references of o is 1
+assert size of o(in bytes) is 4
+assert type of the fields in o is [value]
+```
+
+### 7: Objects as fields
+
+**Case:**
+```
+class Link(object):
+    val: int = 0
+    next: Link = None
+    def add(l: Link) -> Link:
+        m: Link = None
+        m = Link()
+        l.next = m
+        return m
+
+x: Link = None
+y: Link = None
+y = x.add()
+```
+**Expected:**
+
+Let `o` be the object referred to by variable `x` <br>
+Let `p` be the object referred to by variable `y`
+```
+assert number of references of o is 1
+assert type of fields in o is [value, pointer]
+assert number of references of p is 2
+assert type of fields in p is [value, pointer]
+```
+
+### 8. Anonymous object deletion
+
+**Case:**
+```
+class Link(object):
+    val: int = 0
+    next: Link = None
+    def add(l: Link) -> Link:
+        l.next = Link()
+        return l.next
+
+x: Link = None
+x.add()
+x = None
+```
+**Expected:**
+```
+assert number of references of any object is 0
+```
+
+### 9. Simple linked cycle
+**Case:**
+```
+class Link(object):
+    val: int = 0
+    next: Link = None
+
+x: Link = None
+y: Link = None
+x = Link()
+y = Link()
+x.next = y
+y.next = x
+```
+**Expected:**
+
+Let `o` be the object referred to by variable `x` <br>
+Let `p` be the object referred to by variable `y`
+```
+assert number of references of o is 1
+assert type of fields in o is [value, pointer]
+assert number of references of p is 1
+assert type of fields in p is [value, pointer]
+```
+
+### 10. Simple deletion in cycle
+**Case:**
+```
+class Link(object):
+    val: int = 0
+    next: Link = None
+
+x: Link = None
+y: Link = None
+x = Link()
+y = Link()
+x.next = y
+y.next = x
+
+x = None
+```
+**Expected:**
+
+Let `o` be the object referred to by variable `x` <br>
+Let `p` be the object referred to by variable `y`
+```
+assert number of references of o is 0
+assert type of fields in o is [value, pointer]
+assert number of references of p is 1
+assert type of fields in p is [value, pointer]
+```
+
+## Changes to IR
+
+- ### Addition of `ref` value
+```
+export type Value<A> =
+...
+| { a?: A, tag: "ref"}
+```
+Currently there is no way of differentiating between pointers and values stored in an object. As seen in test case *8: Anonymous object deletion*, we need to follow references stored in an object recursively so that we can update their respective reference counts. This is not possible without knowing which field represents a reference/pointer.
+
+- ### Changes to `alloc` expression
+```
+export type Expr<A> =
+  ...
+  | {  a?: A, tag: "alloc", amount: Value<A>[] }
```

Comment:
So it looks like the change here is that this will allocate a fixed list of values (always). Is `amount` the right name for it, then?

Is the idea that in code generation you'll give some kind of information to the `alloc` function about whether each of these is a `ref` or not? How will the code generation work here, because the `alloc` function can't be variable-arity.

I definitely agree that you need some type information to represent this to alloc. Could you make use of the `a` field to give you the type of the contents, and set the metadata that way?

Here's the case I'm concerned about – the `lists` group presumably will want to implement some kind of array list. This involves creating fixed-length arrays with computed sizes and moving elements around. That means the _number_ of elements isn't statically known. How should that group accomplish their task with only fixed-length alloc?

Reply:
From our conversation today, I understand that we can look up the type of an object by looking at the `a` field in `alloc` and then determining if a field is a `reference` or a `value` by looking up the class information on the fields.


`amount` field can then be used to denote the number of elements of type `a` which are to be allocated in a contiguous fashion on the memory. This would also solve the issue with dynamic-size allocation. I think this is what the original IR was going for and we might have misinterpreted some stuff.

I guess we could keep the IR same as original in this case, and get all info we need as below
- Type info - determined from class `fields`
- Size info - determined by the `amount` field
- Field info(reference or value) - determined from class `fields`, since `VarInit` has type info